api/src/main/java/org/apache/iceberg/io/SupportsRecoveryOperations.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+/**
+ * This interface is intended as an extension for FileIO implementations to provide additional
+ * best-effort recovery operations that can be useful for repairing corrupted tables where there are
+ * reachable files missing from disk. (e.g. a live manifest points to data file entry which no
+ * longer exists on disk)
```

Comment:
The example you mention (manifest points to non existing data files) isn't specific to a `FileIO` implementation, but each `FileIO` implementation would have to implement it. Such a scenario can be handled in places that know how to deal with manifests, manifest-lists, data-files.

Reply:
The issue isn't specific to the `FileIO`, but the remediation is specific to the `FileIO` and each requires a specific implementation.

Reply:
@snazy basically what Dan said, for example S3FileIO would have it's own implementation for remediation by going through s3 object versions, same principle for GCS/Azure/HDFS. The example provided is providing an example from the perspective of a user of the interfcace, like a repair function.

For example, with this interface:

```
List<String> missingDataFiles = determineMissingDataFiles(manifest);
successfullyRecoveredFiles = new AtomicInteger(0);
if (fileIO instanceOf SupportsRecoveryOperations) {
missingDataFiles.foreach(missingFile -> {
   if (fileIO.recoverFile(missingFile) {
        successfullyRecoveredFiles.getAndIncrement();
   }
}
} 
```

Reply:
Okay, can you clarify the javadoc saying that the function's purpose is to do "its best" to recover a single file/object. It can be misinterpreted in a way that one can pass in a manifest-file to repair/recover data files referenced from that manifest-file.

Reply:
Sure that sounds reasonable to me!